crates/ra_hir/src/nameres/collector.rs

```diff
@@ -42,14 +42,40 @@ pub(super) fn collect_defs(db: &impl DefDatabase, mut def_map: CrateDefMap) -> C
         unresolved_imports: Vec::new(),
         unexpanded_macros: Vec::new(),
         global_macro_scope: FxHashMap::default(),
-        marco_stack_count: 0,
+        macro_stack_monitor: SimpleMacroStackMonitor::default(),
     };
     collector.collect();
     collector.finish()
 }
 
+trait MacroStackMonitor {
+    fn increase(&mut self, macro_def_id: MacroDefId);
+    fn decrease(&mut self, macro_def_id: MacroDefId);
+
+    fn is_poison(&self, macro_def_id: MacroDefId) -> bool;
+}
```

Comment:
Why do we need a trait for this? Could a concrete data type work for this?

Reply:
It is because we want to allow add a mock monitor for that. We want to assert the count and allow to control the poison validation in test. I can add some test only data in that monitor instead of using trait if you prefer,

Reply:
Ah, I see....

Yeah, I think it's better to add test-only data to concrete type. If it causes problems, we can `#[cfg(test)]` it away